File: 61-rotate-list/rotate-list.py
class Solution:
    def rotateRight(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:

        if head==None or head.next==None or k==0:
            return head
        # 1st step

        crr=head
        length=0
        while crr!=None:
            length+=1
            crr=crr.next

        k=k%length
        

        c=1
        fast=head
        while c<=k:
            c+=1
            fast=fast.next
        
        slow=head
        if fast==None: #length and rotations are same so zero rot
            return head

        while fast.next!=None:
            slow=slow.next
            fast=fast.next

        fast.next=head
        rot_head=slow.next
        slow.next=None


        return rot_head

File: 61-rotate-list/rotate-list.c
struct ListNode* rotateRight(struct ListNode* head, int k) {
    
    if(head==NULL || head->next==NULL){
        return head;
    }
    //1. length
    struct ListNode *crr=head;

    int length=1;
    while(crr->next!=NULL){
        crr=crr->next;
        length++;
    }
    
    //2. k=k%len
    k=k%length;

    //3. tail->head
    crr->next=head;

    //4. 
    int pos=length-k;
    crr=head;
    int count=1;
    while(count<pos){
        crr=crr->next;
        count++;
    }
    //5. 
    head=crr->next;

    //6.
    crr->next=NULL;


    return head;

}